### PR TITLE
refactor (release): refactor the actions release script to match the …

### PR DESCRIPTION
**Pull request by jLopezbarb**
…new release process

# Proposed changes

Fixes DEV-435

This PR introduces a comprehensive script to automate the release process of our GitHub Actions in alignment with stable Okteto CLI releases. The script is designed to:

- Modify the `Dockerfile` in each Action repository to use the specified Okteto CLI version.
- Create or update tags (`vMAJOR`, `vMAJOR.MINOR`, `latest`, `stable`) based on the new release.
- Create a release branch (`release-MAJOR.MINOR`) for targeted fixes without affecting the main branch.
- Include a `--dry-run` option to safely test the release process without making any changes.
- Enhance debugging in case the script fails whenever we execute it 

## How to validate

Please provide step-by-step instructions to replicate your validation scenario. For bug fixes, detail how to reproduce both the bug and its fix, along with any observations.

1.Ensure the GITHUB_TOKEN environment variable is set and has the necessary permissions to push commits and tags and Install all required commands: `semver`, `gsutil`, `git`
1. Comment the following lines:
```sh
if ! grep -qFx "$RELEASE_TAG" "$VERSIONS_FILE"; then
    echo "Error: RELEASE_TAG '$RELEASE_TAG' not found in versions list."
    exit 1
fi
```
3. Run the script: `scripts/ci/new-release-github-actions.sh --dry-run 3.0.0` and check the output

## CLI Quality Reminders 🔧

For both authors and reviewers:

- Scrutinize for potential regressions
- Ensure key automated tests are in place
- Build the CLI and test using the validation steps
- Assess Developer Experience impact (log messages, performances, etc)
- If too broad, consider breaking into smaller PRs
- Adhere to our [code style](https://github.com/okteto/okteto/blob/master/docs/code-style.md) and [code review](https://github.com/okteto/okteto/blob/master/docs/code-review.md) guidelines

### CODE REVIEW

**Review by teresaromero — APPROVED** [review]

**Comment by teresaromero** [comment]
minor: new-release-github-actions.sh can we call it directly release-gh or shorter? would the other one be deprecated? or we keep it for perhaps backports on 2.x versions?

File: scripts/ci/push-image.sh [hunk 1]
```diff
@@ -26,7 +26,7 @@
         beta_prerel_regex="^beta\.[0-9]+"
         prerel="$(semver get prerel "${RELEASE_TAG}" || true)"
         version="$(semver get release "${RELEASE_TAG}" || true)"
-        tags="okteto/okteto:${RELEASE_TAG},okteto/okteto:dev"
+        tags="okteto/okteto:${RELEASE_TAG},okteto/okteto:dev,okteto/okteto:master"
```
Comment (ifbyol) [inline scripts/ci/push-image.sh:29]: This looks wrong. `master` should replace the current `latest`, but this is just adding the tag `master`, not replacing anyone

**Review by andreafalzetti — APPROVED** [review]

**Review by ifbyol — COMMENTED** [review]
Let's make the change we talked offline about pushing `dev` tag only on nightlies. If not, `dev` tag might change between majors and minors versions depending on how we do releases or in case there are backports

File: scripts/ci/release-github-actions.sh [hunk 2]
```diff
@@ -1,139 +1,267 @@
...
-                git config user.name "okteto"
-                git config user.email "ci@okteto.com"
-
-                if git tag --list | grep -qFx "$RELEASE_TAG"; then
-                        echo "$RELEASE_TAG already exists"
-                        exit 1
-                fi
+    # Update 'latest' and 'stable' tags if RELEASE_TAG is greater
+    update_tag_if_greater "latest"
+    update_tag_if_greater "stable"
```
Comment (ifbyol) [inline scripts/ci/release-github-actions.sh:207]: This has to be always updated, right? Also, how does `latest` and `stable` compare to a semantic version (which is the operation done within `update_tag_if_greater`)?

**Review by ifbyol — COMMENTED** [review]
are we going to do somethign special for the current version 2? I mean, creating `v2` and `v2.31` tags?

File: scripts/ci/push-image.sh [hunk 3]
```diff
@@ -17,29 +17,68 @@
...
+                prerel="$(semver get prerel "${tag}" || true)"
+                version="$(semver get release "${tag}" || true)"
+                echo "  DEBUG: prerel: ${prerel}"
+                echo "  DEBUG: version: ${version}"
 
-        beta_prerel_regex="^beta\.[0-9]+"
-        prerel="$(semver get prerel "${RELEASE_TAG}" || true)"
-        version="$(semver get release "${RELEASE_TAG}" || true)"
-        tags="okteto/okteto:${RELEASE_TAG},okteto/okteto:dev"
+                tags_array+=("jlopezbarb/okteto:${tag}")
```
Comment (ifbyol) [inline scripts/ci/push-image.sh:48]:
```suggestion
                tags_array+=("okteto/okteto:${tag}")
```

?

File: scripts/ci/push-image.sh [hunk 4]
```diff
@@ -17,29 +17,68 @@
...
 
-        # if release tag is  not empty, push the stable image
-        if [ -n "$RELEASE_TAG" ]; then
-                echo "prerel: ${prerel}"
-                echo "version ${version}"
-                if [[ $prerel =~ $beta_prerel_regex ]]; then
-                        tags="${tags},okteto/okteto:beta"
+                if [ -n "$prerel" ]; then
+                        if [ "$beta_added" = false ]; then
+                                tags_array+=("jlopezbarb/okteto:beta")
```
Comment (ifbyol) [inline scripts/ci/push-image.sh:54]:
```suggestion
                                tags_array+=("okteto/okteto:beta")
```

?

File: scripts/ci/push-image.sh [hunk 5]
```diff
@@ -17,29 +17,68 @@
...
+                version="$(semver get release "${tag}" || true)"
+                echo "  DEBUG: prerel: ${prerel}"
+                echo "  DEBUG: version: ${version}"
 
-        beta_prerel_regex="^beta\.[0-9]+"
-        prerel="$(semver get prerel "${RELEASE_TAG}" || true)"
-        version="$(semver get release "${RELEASE_TAG}" || true)"
-        tags="okteto/okteto:${RELEASE_TAG},okteto/okteto:dev"
+                tags_array+=("jlopezbarb/okteto:${tag}")
+                echo "  DEBUG: Added tag to tags_array: jlopezbarb/okteto:${tag}"
```
Comment (ifbyol) [inline scripts/ci/push-image.sh:49]:
```suggestion
                echo "  DEBUG: Added tag to tags_array: okteto/okteto:${tag}"
```

?

File: scripts/ci/push-image.sh [hunk 6]
```diff
@@ -17,29 +17,68 @@
...
+                                beta_added=true
+                                echo "  DEBUG: Added beta tag to tags_array"
+                        else
+                                echo "  DEBUG: Beta tag already added, skipping"
+                        fi
                 elif [ -n "$version" ]; then
-                        tags="${tags},okteto/okteto:stable"
+                        # It's a stable version because it has $version and it's not a prerrelease
+                        if [ "$stable_added" = false ]; then
+                                tags_array+=("jlopezbarb/okteto:stable" "jlopezbarb/okteto:latest")
```
Comment (ifbyol) [inline scripts/ci/push-image.sh:63]:
```suggestion
                                tags_array+=("okteto/okteto:stable" "okteto/okteto:latest")
```

?

File: scripts/ci/release-github-actions.sh (hunk 2 above)
Comment (jLopezbarb) [reply scripts/ci/release-github-actions.sh:207]: We are not comparing the `latest` and `stable` with a semver.

I will try to explain the function `update_tag_if_greater` the best I can. Also I will add some comments in order to make it more explainable:

1. Check if the tag already exists

> if git rev-parse "${TAG_NAME}" >/dev/null 2>&1; then

We check if latest or stable exists as a tag

2. Getting the last official stable version

> current_tag_version=$(git describe --tags "${TAG_NAME}" --abbrev=0 --match "*.*.*" 2>/dev/null || echo "0.0.0")

This command is to retrieve the semver associated with a specific Git tag (TAG_NAME). If it fails to find such a version, it defaults to `0.0.0`. For example in the following screenshot, if you pass latest as argument we return `1.21.3`
Screenshot 2024-10-03 at 13 41 11

3. Compare with the version that we got in previous command:

> diff=$(semver compare "${RELEASE_TAG}" "${current_tag_version}")

We check if the release we are releasing is greater than the one is set as latest.

**Comment by jLopezbarb** [comment]
> are we going to do somethign special for the current version 2? I mean, creating `v2` and `v2.31` tags?

We are going to leave it as it is. `v2` will still be a branch and `2.31` is already a tag

File: scripts/ci/release-github-actions.sh (hunk 2 above)
Comment (ifbyol) [reply scripts/ci/release-github-actions.sh:214]: Ah, Ok, I had missed the command to retrieve the semantic version from the `latest` or `stable` tag, thanks for the explanation

**Pull request merged**